timely/src/dataflow/operators/core/input.rs

```diff
@@ -274,26 +357,48 @@ impl<T: Timestamp, C: Container> Handle<T, C> {
         self.pushers.push(pusher);
     }
 
-    // flushes our buffer at each of the destinations. there can be more than one; clone if needed.
-    #[inline(never)]
+    /// Extract all ready contents from the builder and distribute to downstream operators.
+    #[inline]
+    fn extract(&mut self) {
+        while let Some(container) = self.builder.extract() {
+            Self::flush_container(container, &mut self.buffer, &mut self.pushers, &self.now_at);
+        }
+    }
+
+    /// Flush all contents and distribute to downstream operators.
+    #[inline]
     fn flush(&mut self) {
-        for index in 0 .. self.pushers.len() {
-            if index < self.pushers.len() - 1 {
-                self.buffer2.clone_from(&self.buffer1);
-                Message::push_at(&mut self.buffer2, self.now_at.clone(), &mut self.pushers[index]);
-                debug_assert!(self.buffer2.is_empty());
+        while let Some(container) = self.builder.finish() {
+            Self::flush_container(container, &mut self.buffer, &mut self.pushers, &self.now_at);
+        }
+    }
+
+    /// flushes our buffer at each of the destinations. there can be more than one; clone if needed.
+    /// Does not take `self` because `flush` and `extract` borrow `self` mutably.
+    #[inline(never)]
+    fn flush_container(
```

Comment:
Future TODO: We should rationalize the words used here to e.g. perhaps change this to `push_container` or .. eh .. `extract_container` or something that communicates "move some amount of work out of the handle" more than "be done". Also fix up `close_epoch` similarly.